### PR TITLE
feat(ui5-test-writer): Generate tests for Contact Cards

### PR DESCRIPTION
**Pull request by sap-sebelao**
- covers both LR and OP as there is a lot of shared code anyway
- also contains some minor fixes
-- column availability is now evaluated
-- commented out test suggestion now suggests existing action

Internal issue
36742
Internal JIRAs
FXUBRQ34-12564
FXUBRQ34-12565
FXUBRQ34-12566
FXUBRQ34-12525
FXUBRQ34-12604

### CODE REVIEW

File: packages/ui5-test-writer/templates/v4/integration/pages/ObjectPage.ts [hunk 1]
```diff
@@ -1,14 +1,4 @@
-import type Opa5 from "sap/ui/test/Opa5";
```
Comment (dudajSAP) [inline packages/ui5-test-writer/templates/v4/integration/pages/ObjectPage.ts]: You added in previous PR https://github.com/SAP/open-ux-tools/pull/4811 the auto-generated header into `packages/ui5-test-writer/templates/v4/integration/pages/ListReport.js` and `packages/ui5-test-writer/templates/v4/integration/pages/ObjectPage.js`.

Now I expect the same also for TS-variant:

- `ListReport.ts`
- `ObjectPage.ts`
- `OpaJourneyTypes.d.ts`

I'm sorry, I missed it in the previous PR.

**Review by dudajSAP — CHANGES_REQUESTED** [review]
From my POV now the auto-gen header is missing on multiple places.

File: packages/ui5-test-writer/templates/v4/integration/pages/ObjectPage.ts (hunk 1 above)
Comment (sap-sebelao) [reply packages/ui5-test-writer/templates/v4/integration/pages/ObjectPage.ts]: Good catch, I forgot to add the warnings already during #4811 

Will add them.

File: packages/ui5-test-writer/src/utils/objectPageUtils.ts [hunk 2]
```diff
@@ -329,10 +374,21 @@ function extractFormFields(subSection: BodySectionItem): SectionFormField[] {
...
+
+        if (annotationParts) {
+            fields.push({
+                property: `${baseProperty}/${annotationParts.targetAnnotation}`,
+                targetAnnotation: annotationParts.targetAnnotation
+            });
+        } else {
+            fields.push({ property: baseProperty });
         }
     });
```
[the remark this answers is not in the document]
Comment (sap-sebelao) [reply packages/ui5-test-writer/src/utils/objectPageUtils.ts:542]: The Contact-card iCheckField({ property: "_Customer/Contact" }) call actually passes at runtime, it's intentional dual diagnostic alongside iClickLink + iCheckContactDialog (separates "field rendered" from "link interacts").

However, investigating this issue uncovered real runtime failures with ConnectedFields and FieldGroup wrappers.

Those need proper handling via the connectedFields / fieldGroup keys on FieldIdentifier.

Since the fix is a bit larger and out of scope for this Contact-cards PR, it will be handled in a separate PR.

On this branch the wrappers are now silently skipped so they no longer produce broken assertions, as a temporary workaround.

File: packages/ui5-test-writer/src/utils/objectPageUtils.ts [hunk 3]
```diff
@@ -298,23 +308,58 @@ function extractSectionActions(
...
         subSections.push({
             id: subSectionId,
             navigationProperty: getNavigationPropertyFromKey(subSectionKey),
             isTable: !!subSection.isTable,
             custom: !!subSection.custom,
             order: subSection?.order ?? -1, // put a negative order number to signal that order was not in spec
-            fields: subSection.custom || subSection.isTable ? [] : extractFormFields(subSection),
-            tableColumns: subSection.custom || !subSection.isTable ? {} : extractTableColumnsFromNode(subSection)
+            fields,
+            contactCardFields: pickContactCardFields(fields),
```
[the remark this answers is not in the document]
Comment (sap-sebelao) [reply packages/ui5-test-writer/src/utils/objectPageUtils.ts]: This is intentional, see first reply.

File: packages/ui5-test-writer/templates/v4/integration/ObjectPageJourney.js [hunk 4]
```diff
@@ -120,26 +120,42 @@ sap.ui.define([
...
+<% }) -%>
 <% } -%>
 <% if (section?.subSections?.length > 0) { -%>
 <% section.subSections.forEach(function(subSection) { -%>
-            //When.onThe<%- name%>Generated.iGoToSection({ section: "<%- section.id %>", subSection: "<%- subSection.id %>" });
+            When.onThe<%- name%>Generated.iGoToSection({ section: "<%- section.id %>", subSection: "<%- subSection.id %>" });
             Then.onThe<%- name%>Generated.iCheckSubSection({ section: "<%- subSection.id %>" });
 <% if (subSection.fields && subSection.fields.length > 0) { -%>
 <% subSection.fields.forEach(function(field) { -%>
             Then.onThe<%- name%>Generated.onForm({ section: "<%- subSection.id %>" }).iCheckField({ property: "<%- field.property %>" });
```
[the remark this answers is not in the document]
Comment (sap-sebelao) [reply packages/ui5-test-writer/templates/v4/integration/ObjectPageJourney.js]: This is intentional, see first reply.

File: packages/ui5-test-writer/src/utils/objectPageUtils.ts [hunk 5]
```diff
@@ -329,10 +374,25 @@ function extractFormFields(subSection: BodySectionItem): SectionFormField[] {
...
+        const annotationParts = parseDataFieldForAnnotationName(fieldItem.name);
+        const valueProperty = fieldItem.schema?.keys?.find((key) => key.name === 'Value')?.value;
+        const baseProperty = valueProperty ?? annotationParts?.property;
+        if (!baseProperty) {
+            return;
+        }
+
+        if (annotationParts) {
+            if (annotationParts.targetAnnotation !== 'Contact') {
+                // TODO: Handle other annotation-typed wrappers (such as ConnectedFields, FieldGroup)
```
Comment (sap-sebelao) [inline packages/ui5-test-writer/src/utils/objectPageUtils.ts]: Followup for this TODO: https://github.com/SAP/open-ux-tools/pull/4878

File: packages/ui5-test-writer/templates/v4/integration/pages/ObjectPage.ts (hunk 1 above)
Comment (sap-sebelao) [reply packages/ui5-test-writer/templates/v4/integration/pages/ObjectPage.ts]: Added in https://github.com/SAP/open-ux-tools/pull/4879, keeping it separate so that the TS banners have a chance to be added before dev close.

**Comment by sap-sebelao** [comment]
On a second thought - to offload some fixes off of this PR, to keep it cleaner, I moved the `iGoToSection` workaround removal to another branch/PR: https://github.com/SAP/open-ux-tools/pull/4881

I will shortly remove the same changes from this one, to clean it up and make it focused only on the Contact Cards + minor fixes that would otherwise block the Contact Card tests.

**Review by dudajSAP — DISMISSED** [review]

**Comment by sap-sebelao** [comment]
The folllow up fix https://github.com/SAP/open-ux-tools/pull/4878 was rebased to main to be able to merge the fix before this feature PR.

This will result in a conflict when https://github.com/SAP/open-ux-tools/pull/4878 is merged.

Draft mode for now - waiting for https://github.com/SAP/open-ux-tools/pull/4878

File: packages/ui5-test-writer/templates/v4/1.150/integration/ObjectPageJourney.js [hunk 6]
```diff
@@ -130,16 +138,28 @@ sap.ui.define([
...
 <% }) -%>
 <% } -%>
+<% subSection.contactCardFields.forEach(function(field) { -%>
+            When.onThe<%- name%>Generated.onForm({ section: "<%- subSection.id %>" }).iClickLink({ property: "<%- field.property %>" });
+            Then.onThe<%- name%>Generated.onDialog().iCheckContactDialog({ controlType: "sap.ui.mdc.link.Panel" });
+<% }) -%>
 <% if (subSection.tableColumns && Object.keys(subSection.tableColumns).length > 0 && subSection.navigationProperty) { -%>
             Then.onThe<%- name%>Generated.onTable({ property: "<%- subSection.navigationProperty %>" }).iCheckColumns(<%- JSON.stringify(subSection.tableColumns) %>);
 <% } -%>
+<% subSection.contactCardColumns.forEach(function(column) { -%>
```
Comment (heimwege) [inline packages/ui5-test-writer/templates/v4/1.150/integration/ObjectPageJourney.js:160]: (and the equivalent line in ObjectPageJourney.ts). Same guard as line 145 (`iCheckColumns`) is missing here. Because onTable({ property: ... }) requires a navigation property to identify which table to act on. If navigationProperty is undefined, the generated test call is meaningless - there's no table to target.

  The iCheckColumns block already has this guard for exactly the same reason. The contactCardColumns block does the same onTable({ property: "<%- subSection.navigationProperty %>" }) call, so it needs the same guard.

**Review by heimwege — COMMENTED** [review]
see comment

File: packages/ui5-test-writer/templates/v4/1.150/integration/ObjectPageJourney.js (hunk 6 above)
Comment (sap-sebelao) [reply packages/ui5-test-writer/templates/v4/1.150/integration/ObjectPageJourney.js:160]: Fixed in https://github.com/SAP/open-ux-tools/pull/4872/changes/bf100cc00a46c5acf1510d94c84cbe7e54e35b66

File: packages/ui5-test-writer/src/utils/objectPageUtils.ts (hunk 5 above)
Comment (devinea) [reply packages/ui5-test-writer/src/utils/objectPageUtils.ts]: https://github.com/SAP/open-ux-tools/pull/4878 is merged.

File: packages/ui5-test-writer/templates/v4/1.150/integration/ObjectPageJourney.js [hunk 7]
```diff
@@ -120,26 +124,44 @@ sap.ui.define([
             Then.onThe<%- name%>Generated.onTable({ property: "<%- section.navigationProperty %>" }).iCheckDelete({ visible: true });
             // When.onThe<%- name%>Generated.onTable({ property: "<%- section.navigationProperty %>" }).iPressDelete();
 <% } -%>
+<% section.contactCardColumns.forEach(function(column) { -%>
+            When.onThe<%- name%>Generated.onTable({ property: "<%- section.navigationProperty %>" }).iClickLink(0, "<%- column.property %>");
+            Then.onThe<%- name%>Generated.onDialog().iCheckContactDialog({ controlType: "sap.ui.mdc.link.Panel" });
+<% }) -%>
```
[the remark this answers is not in the document]
Comment (sap-sebelao) [reply packages/ui5-test-writer/templates/v4/1.150/integration/ObjectPageJourney.js:141]: Already guarded. The outer `if (section.isTable && section.navigationProperty) (js:120 / ts:139)` wraps this contactCardColumns loop, so navigationProperty is guaranteed defined here.

File: packages/ui5-test-writer/templates/v4/1.150/integration/ObjectPageJourney.ts [hunk 8]
```diff
@@ -139,26 +143,44 @@ function journey() {
         Then.onThe<%- name%>Generated.onTable({ property: "<%- section.navigationProperty %>" }).iCheckDelete({ visible: true });
         // When.onThe<%- name%>Generated.onTable({ property: "<%- section.navigationProperty %>" }).iPressDelete();
 <% } -%>
+<% section.contactCardColumns.forEach(function(column) { -%>
+        When.onThe<%- name%>Generated.onTable({ property: "<%- section.navigationProperty %>" }).iClickLink(0, "<%- column.property %>");
+        Then.onThe<%- name%>Generated.onDialog().iCheckContactDialog({ controlType: "sap.ui.mdc.link.Panel" });
+<% }) -%>
```
[the remark this answers is not in the document]
Comment (sap-sebelao) [reply packages/ui5-test-writer/templates/v4/1.150/integration/ObjectPageJourney.ts:163]: Already guarded. The outer `if (section.isTable && section.navigationProperty) (js:120 / ts:139)` wraps this contactCardColumns loop, so navigationProperty is guaranteed defined here.

File: packages/ui5-test-writer/src/utils/objectPageUtils.ts [hunk 9]
```diff
@@ -295,51 +307,169 @@ function extractSectionActions(
...
+function extractBodySubSectionsData(
+    section: SectionItem,
+    parentSectionId: string,
+    convertedMetadata?: ConvertedMetadata,
+    entitySetName?: string
+): BodySubSectionFeatureData[] {
     const subSections: BodySubSectionFeatureData[] = [];
-    const subSectionsAggregation = getAggregations(section)['subSections'];
+    const sectionAggregations = getAggregations(section);
+    const subSectionsAggregation = sectionAggregations['subSections'] ?? sectionAggregations['subsections'];
```
[the remark this answers is not in the document]
Comment (sap-sebelao) [reply packages/ui5-test-writer/src/utils/objectPageUtils.ts]: Original implementation used lowercase `subsections` and there were no issues raised related to that.

However, during implementing Contact Cards, a version of spec model using `subSections` came up, so this is set up to work with both eventualities.

The test is actually there not sure what the AI is complaining about.

File: packages/ui5-test-writer/test/unit/utils/modelUtils.test.ts [hunk 10]
```diff
@@ -369,3 +401,36 @@ describe('Test edge cases for better branch coverage', () => {
...
+    test('returns undefined for falsy input', () => {
+        expect(parseDataFieldForAnnotationName(undefined)).toBeUndefined();
+        expect(parseDataFieldForAnnotationName('')).toBeUndefined();
+    });
+
+    test('returns undefined when property or annotation segment is empty', () => {
+        expect(parseDataFieldForAnnotationName('DataFieldForAnnotation::::Contact')).toBeUndefined();
+        expect(parseDataFieldForAnnotationName('DataFieldForAnnotation::Customer::')).toBeUndefined();
+    });
+});
```
[the remark this answers is not in the document]
Comment (sap-sebelao) [reply packages/ui5-test-writer/test/unit/utils/modelUtils.test.ts]: Fixed - tests block were not entirely identical, but similar enough to be consolidated into 1.

File: packages/ui5-test-writer/src/utils/objectPageUtils.ts [hunk 11]
```diff
@@ -295,51 +307,169 @@ function extractSectionActions(
...
+/**
+ * Filters form fields down to those rendered as Contact-card links (`@Communication.Contact`).
+ *
+ * @param fields - all form fields of a (sub-)section
+ * @returns Contact-card fields, addressed via the qualified `<property>/<targetAnnotation>` form
+ */
+function pickContactCardFields(fields: SectionFormField[]): ContactCardField[] {
+    return fields
+        .filter((field) => field.targetAnnotation === 'Contact')
+        .map((field) => ({ property: field.property }));
```
[the remark this answers is not in the document]
Comment (sap-sebelao) [reply packages/ui5-test-writer/src/utils/objectPageUtils.ts:472]: Comment added at line of assignment

File: packages/ui5-test-writer/src/types.ts [hunk 12]
```diff
@@ -127,6 +127,7 @@ export type SectionFormField = {
     property: string;
     connectedFields?: string;
     fieldGroup?: string;
+    targetAnnotation?: string;
 };
 
 export type TableColumn = {
```
Comment (heimwege) [inline packages/ui5-test-writer/src/types.ts:140]: Do we need a new type def for this one? Isn't that `type sap.ui.mdc.table.Column`?
Same question for `ContactCardField`. Shouldn't there be an existing type form the UI5 types?

Maybe not in this PR but we can have a cleanup and get rid of redundant types and use/enhance the existing ones from the UI5 controls

**Review by heimwege — DISMISSED** [review]
- changeset ok
- review comments addressed
- coverage is excellent
- did NOT test manually

Added one potential follow-up activity

File: packages/ui5-test-writer/src/types.ts (hunk 12 above)
Comment (sap-sebelao) [reply packages/ui5-test-writer/src/types.ts:140]: These types aren't supposed to be representations of the actual UI5 controls or supposed to contain all information about them, using the full type from UI5 would probably result in more problems.
These are a description of an internal object of the generator, that carries only the relevant information for generating test code in the EJS templates.

I can see where the confusion comes from, the naming is not very helpful - It should probably be something along the lines of "TableColumnInformation"/"TableColumnData"/"TableColumnProperties"...

A follow up for better naming would make sense.

File: .changeset/ui5-test-writer-contact-cards.md [hunk 13]
```diff
@@ -0,0 +1,5 @@
+---
+"@sap-ux/ui5-test-writer": minor
+---
+
+FEAT: Generate Contact Card OPA5 tests across Object Page header field groups, body-section forms, body-section tables, and List Report tables. `DataFieldForAnnotation::<property>::Contact` entries are detected in the spec model and emitted as `iClickLink({ property: "<property>/Contact" })` followed by `iCheckContactDialog({ controlType: "sap.ui.mdc.link.Panel" })`. Drops the `iPressSectionIconTabFilterButton` page-object workaround in favor of the public `iGoToSection` API and unconditionally emits `iCheckNumberOfSections` (now valid for any section count).
```
Comment (docirl) [inline .changeset/ui5-test-writer-contact-cards.md:5]: There are 2 changesets, please check which one is needed or merge

Comment (sap-sebelao) [reply .changeset/ui5-test-writer-contact-cards.md:5]: Removed the extra ones.

It looks like AI is interpreting the AGENTS.md in a way where it has to create a changeset on specific code changes, but disregards if a changeset is already present. But only sometimes.

My fault for not noticing that though.

File: packages/ui5-test-writer/src/types.ts (hunk 12 above)
Comment (heimwege) [reply packages/ui5-test-writer/src/types.ts:140]: If we just need specific properties from them (and enhance with our internal stuff) it might still be worth `pick`ing from the original type to avoiding diverging types in case UI5 does adjustments.

**Review by heimwege — DISMISSED** [review]
re-approve

File: packages/ui5-test-writer/templates/v4/integration/ObjectPageJourney.ts [hunk 14]
```diff
@@ -81,7 +81,7 @@ function journey() {
 <% } -%>
 
 <% if (headerSections?.length > 0) { -%>
-    opaTest("Check header facets of the Object Page", function (_Given: Given, _When: When, Then: Then) {
+    opaTest("Check header facets of the Object Page", function (_Given: Given, When: When, Then: Then) {
```
Comment (docirl) [inline packages/ui5-test-writer/templates/v4/integration/ObjectPageJourney.ts]: ### Unused `When` parameter in generated header-facets OPA test

**File:** `packages/ui5-test-writer/templates/v4/integration/ObjectPageJourney.ts:82`

The "Check header facets" `opaTest` signature was changed from `_When: When` to `When: When`, but the only `When.` usage in that block is gated on `section.contactCardFields.forEach(...)`:

```ts
opaTest("Check header facets of the Object Page", function (_Given: Given, When: When, Then: Then) {
```

Every other emitted line in the block uses `Then.` (`iCheckMicroChart`, `iCheckHeaderFacet`, …). The same template already toggles the underscore prefix based on usage elsewhere:

```ts
function (_Given: Given, <% if (bodySections?.length > 1) { %>When: When<% } else { %>_When: When<% } %>, Then: Then)
```

**Failure scenario:** An Object Page with header facets but no `@Communication.Contact` header fields (e.g. a microchart facet, or a header form with only non-Contact fields) emits this test with an unreferenced `When` parameter. The header-facets test is generated whenever `headerSections?.length > 0`, independent of contact fields, so this is a common case. The resulting `.gen.ts` fails with **TS6133 "'When' is declared but its value is never read"** under `noUnusedParameters` / `@typescript-eslint/no-unused-vars`.

The `.js` template is unaffected (no type parameters).

**Suggested fix:** Gate the prefix the same way the body-section block does:

```ts
function (_Given: Given, <% if (headerSections.some(function(s) { return s.contactCardFields.length > 0; })) { %>When: When<% } else { %>_When: When<% } %>, Then: Then) {
```

Comment (sap-sebelao) [reply packages/ui5-test-writer/templates/v4/integration/ObjectPageJourney.ts]: Fixed in https://github.com/SAP/open-ux-tools/pull/4872/commits/8ce436df58aedf748297a3cc86bfc75330347dd1

**Review by heimwege — DISMISSED** [review]
re-approve

**Review by devinea — DISMISSED** [review]
Code changes seem clear and well covered by tests. 
did not test locally.

**Review by devinea — APPROVED** [review]
Re-approving after merge conflicts fixed.
